### ISSUE TITLE
Best practice for using max 2 cores on CRAN?

### ISSUE THREAD

**Issue by tdhock** [Rdatatable/data.table#5658]
CRAN has a policy that limits the number of cores that can be used during package checks, and they have recently implemented a mechanism for checking compliance. Upon recent CRAN submission of one of my R packages which imports data.table, I got the following rejection message from CRAN:
```
Flavor: r-devel-linux-x86_64-debian-gcc
Check: examples, Result: NOTE
  Examples with CPU (user + system) or elapsed time > 5s
                         user system elapsed
  aum_line_search      12.349  0.322   1.935
  aum_line_search_grid 10.033  0.308   1.781
  Examples with CPU time > 2.5 times elapsed time
                         user system elapsed ratio
  aum_line_search      12.349  0.322   1.935 6.548
  aum_line_search_grid 10.033  0.308   1.781 5.806
  aum_diffs_penalty     4.730  0.169   1.635 2.996
```
These messages can be suppressed by adding data.table::setDTthreads(1) at the start of each example.
Is there another/recommended way to avoid using too many cores when checking a CRAN package which uses data.table? The default number of threads in data.table is max cores/2 which is over the CRAN limit (max 2 cores I believe).
Also is there documentation about this somewhere? Probably would be good to mention in the datatable-importing vignette.

**Comment by eddelbuettel** [comment]
> they have recently implemented a mechanism for checking compliance

I don't think it is that recent. Package `tiledb` got dinged three or so years ago, and I then implemented a scheme that works.  

Here is what the package does now.  In essence, in every example file or unit test file (i.e. files that CRAN runs) I _explicitly_ call a function that throttles cores down to 
- the given function value, if there is one, in this throttler function
- or the value of option `Ncpu` with a fallback value of 2 (and that is what CRAN gets; as I recall they set `Ncpu`)
- or the value of environment variable `OMP_THREAD_LIMIT` common to set OpenMP and MKL thread counts.

So this gets us the max of 2 _in everything that CRAN runs_ yet it can be overriden (if, they, your CI has more cores).  I already had `Ncpu` set locally to e.g. install multiple packages in parallel via `install.packages()` / `update.package()` ).  

Importantly, it leaves general startup (i.e. `.onLoad()`, `.onAttach()` ) alone: it still gets full core count and performance. 

So normal users are not affected and get full speed, but CRAN gets its limit.  I can detail that some more you if you care; it is in `tiledb` file `R/config.R` (but a little tied to how the config object progagates for us).

**Comment by jangorecki** [comment]
Also reading DT news file can point to commits that set that up for DT

**Comment by tdhock** [comment]
@eddelbuettel thanks for the quick feedback, that approach seems similar to what I did, but more flexible. You wrote that CRAN sets option Ncpu -- do you know where is that documented? On https://cran.r-project.org/web/packages/policies.html I see "If running a package uses multiple threads/cores it must never use more than two simultaneously: the check farm is a shared resource and will typically be running many checks simultaneously. " but no mention of option Ncpu. 

Also I was thinking that if CRAN indeed sets option Ncpu then the data.table default number of threads should respect that (avoids having to change example/test code in 1000+ packages with hard dependency on data.table).  

@jangorecki I checked https://github.com/Rdatatable/data.table/blob/master/NEWS.md but I did not see mention of commits, can you please clarify?

**Comment by tdhock** [comment]
An alternative to option Ncpu would be to just detect if on CRAN, for example using code below, and then throttle to two cores for data.table default.
```r
> testthat:::on_cran
function () 
{
    !env_var_is_true("NOT_CRAN")
}
```

**Comment by eddelbuettel** [comment]
`NOT_CRAN` is hack I do not recommended.  

I do not recall where they document but it is documented somewhere that they allow two threads.

**Comment by tdhock** [comment]
There is a related R-devel thread, https://stat.ethz.ch/pipermail/r-devel/2021-November/081289.html that explains the env var _R_CHECK_LIMIT_CORES_=TRUE is set when running R CMD check --as-cran.

**Comment by tdhock** [comment]
?tools::check_packages_in_dir says Ncpus option is used to specify how many packages are checked in parallel, 
```
   Ncpus: the number of parallel processes to use for parallel
          installation and checking.
```

**Comment by eddelbuettel** [comment]
But note that what you quoted does not limit it to _package checks_ as your statement implies but to general 'checking'.  Which is where we started: how to play nice at CRAN and not exceed limits.

**Comment by tdhock** [comment]
This check is implemented via this code https://github.com/wch/r-source/blob/1c0545ba7c6c07e8c358eda552b875b1e4d6826d/src/library/tools/R/check.R#L4123 which gets the 2.5 ratio from the environment variable `_R_CHECK_EXAMPLE_TIMING_CPU_TO_ELAPSED_THRESHOLD_` so we could check to see if that is set and then round down as a default for number of threads.

**Comment by eddelbuettel** [comment]
Also: 

https://github.com/Rdatatable/data.table/blob/88039186915028ab3c93ccfd8e22c0d1c3534b1a/src/openmp-utils.c#L95

Recall that my approach was about taking the lower limit from Ncpu amd OMP_NUM_THREADS (and I no longer recall why I do not / did not also set OMP_THREAD_LIMIT).

**Comment by tdhock** [comment]
related to #5573 #5620 about other environment variables to look at to determine default number of threads.

**Comment by jangorecki** [comment]
I haven't found in NEWS file anything useful as well.
https://github.com/Rdatatable/data.table/issues/3300 seems to be related

**Comment by tdhock** [comment]
#3300 quotes an email from Brian Ripley saying that CRAN sets OMP_THREAD_LIMIT=2, but that was back in 2019, so I suspect that @TimTaylor is right that CRAN is no longer doing that, but they are clearly setting `_R_CHECK_EXAMPLE_TIMING_CPU_TO_ELAPSED_THRESHOLD_=2.5` because that is the source of the NOTE in my original post, so I would suggest using this env var instead.

**Comment by tdhock** [comment]
https://github.com/Rdatatable/data.table/issues/3300#issuecomment-1635680140 quotes Uwe Ligges saying that they no longer set the `OMP_THREAD_LIMIT` env var, so I suggest we instead look at `_R_CHECK_EXAMPLE_TIMING_CPU_TO_ELAPSED_THRESHOLD_` and round down.

**Comment by eddelbuettel** [comment]
Your call between a second-hand quote (!!) in a 2019 ticket (!!) on one hand and the [CRAN Repository Policy](https://cran.r-project.org/web/packages/policies.html) on the other. 

The latter still says

> If running a package uses multiple threads/cores it must never use more than two simultaneously: the check farm is a shared resource and will typically be running many checks simultaneously.
>
> Examples should run for no more than a few seconds each: they are intended to exemplify to the would-be user how to use the functions in the package.

**Comment by lrberge** [comment]
Is there something against placing:

```R
if(any(grepl("_R_CHECK", names(Sys.getenv()), fixed = TRUE))){
  setDTthreads(2)
}
```

in `data.table`'s [`onLoad`](https://github.com/Rdatatable/data.table/blob/88039186915028ab3c93ccfd8e22c0d1c3534b1a/R/onLoad.R#L17)?

 It would save all dependent package maintainers to do that down the road. But maybe I'm missing something!

**Comment by eddelbuettel** [comment]
@lrberge That is pretty clever but it would bite folks like me who have values in a dotfile for `R CMD check`:

```r
> grep("_R_", names(Sys.getenv()), value=TRUE)
[1] "_R_CHECK_COMPILATION_FLAGS_KNOWN_" "_R_CHECK_TESTS_NLINES_"           
> 
```

It would be better if we could nudge Uwe towards setting the OMP variable on his machine!  He is the one who wants the lower core and process count there!

**Comment by lrberge** [comment]
@eddelbuettel how come you have "_R_CHECK" env vars lying around?! :-)
I agree the cleanest solution would be the OMP variable.

@TimTaylor: the current default for many packages with multi threading is usually to max out resources (or close to) which in many cases isn't the best. 
So nudging users to set it themselves can be a solution. The big issue is... it makes R very user unfriendly. 

```R
# A)
library(data.table)
setDTthreads(percent = 0.5)

# B)
library(data.table)
```

Running A) instead of B) is really a pain and nobody will do it: this will follow that R multi-core-disabled functions will look slow on large tasks as compared to other non-R software.
Placing A) in an `.Rprofile` is too much to ask for new R users.
If R takes the single core path, it must be written in bold and large font how to change that (in particular the `.Rprofile` mechanism) in as many places as possible, otherwise new users simply will be disappointed with performance on demanding tasks and won't look further.

So in my view: a) using all available (or many) threads as a default isn't great, b) using single thread mode as default and nudging users to change that has big caveats and isn't really user-firendly. So maybe the way is c)? c) "smart" thread setting, with the number of threads decided using the parameters of the task to achieve? (A bit like the `throttle` argument of `setDTthread` but with more detailed heuristics.)
It seems to me that currently the official R core team position and direction isn't really clear (yet).

**Comment by eddelbuettel** [comment]
>  using all available (or many) threads as a default isn't great

Wit my "work" hat on:  I disagree.  We build software for 'big enough' problems and e.g. our underlying library maxes out by default.  I strongly believe that is the right thing.  (And in our package I put a throttler in as described above).

So I actually hold both views:  I like what `data.table` does as a default and would not want it to default to 2, but as a widely-used package it would also be nice to help other packages using it.

Back to: How do we get Uwe to change his mind?  :grin:

**Comment by aadler** [comment]
As the problem seems to be confined when running tests (maybe examples) what I've done is similar to @eddelbuettel, which is write explicit thread handlers, and then call them at the top of my tinytest files. Wouldn't it be simplest to leave environment alone and simply call `setDTthreads(2L)` in each test file?

**Comment by lrberge** [comment]
Nope, it's in the examples too.

**Comment by BerriJ** [comment]
I'm sorry for spamming a little here. Today, I ran into the same problem with an R package of mine (https://github.com/BerriJ/profoc). This is just using OpenMP, not data.table. So, using `setDTthreads(2)` isn't an option for me.
This will bite many Rcpp Package maintainers.
There is a related comment here: https://github.com/Rdatatable/data.table/issues/3300#issuecomment-1635680140

**Comment by eddelbuettel** [comment]
But setting `OPENMP_THREAD_LIMIT` does help you and it is what I advocate for (over in the mailing list thread).

**Comment by lrberge** [comment]
Using `OMP_THREAD_LIMIT` was the recommended way to limit multi-threading. It's still in the [writing R exts](https://cran.r-project.org/doc/manuals/R-exts.html#OpenMP-support)! 
>One way to limit resources is to limit the overall number of threads available to OpenMP in the R process: this can be done via environment variable OMP_THREAD_LIMIT, where implemented.

I cannot understand why it has been stealthily removed without any notice to the maintainers. 

Literally all packages using openMP are impacted **and** all the packages using them as dependencies!!! Must be 40% of all packages. The CRAN team email inbox must be overflowing by now!

**Comment by aadler** [comment]
> I'm sorry for spamming a little here. Today, I ran into the same problem with an R package of mine (https://github.com/BerriJ/profoc). This is just using OpenMP, not data.table. So, using `setDTthreads(2)` isn't an option for me. This will bite many Rcpp Package maintainers. There is a related comment here: [#3300 (comment)](https://github.com/Rdatatable/data.table/issues/3300#issuecomment-1635680140)

Bit me with Delaporte a few weeks ago. I went the @eddelbuettel and @mattdowle route of writing thread handlers and calling them. While frustrating for the developers, it does serve the purpose of letting users default to MAXIMUM POWER while allowing us to throttle CRAN (figuratively only, I hope).

**Comment by BerriJ** [comment]
For reference, I just solved this by adding `Sys.setenv("OMP_THREAD_LIMIT" = 2)` to the top of `tests/testthat.R` and my `vignettes/profoc.Rmd` vignette (https://github.com/BerriJ/profoc/commit/c19940abbc68ab0ee94a1f1fd8c995f3f39e2973).

So `tests/testthat.R` looks like this now:

```
Sys.setenv("OMP_THREAD_LIMIT" = 2)

library(testthat)
library(profoc)

test_check("profoc")

```

I also had to remove a make flag (https://github.com/BerriJ/profoc/commit/05883f41e62222bfd3a7023bb95969eadc479f54) but I'm not 100% sure if this is related or just coincidence. However, CRAN checks threw a note about the package installation time.

Maybe this will help someone. Thanks a lot for your help.

**Comment by eddelbuettel** [comment]
(The `Makeflags` issue is 100% unrelated to this question or thread.  Just edit it out. I'll delete this nag then.)

**Comment by BerriJ** [comment]
> (The `Makeflags` issue is 100% unrelated to this question or thread.  Just edit it out. I'll delete this nag then. Also, there is no NOTE about package installation time. Some packages take 20+ minutes to install)

I got this though ![Screenshot_20230825-233343.png](https://github.com/Rdatatable/data.table/assets/37799358/acdbb5bb-4833-4ecb-8734-85a9e5076b1c)

**Comment by eddelbuettel** [comment]
Ohhhh.  I stand corrected.  That is indeed the same 'too much parallelism', I had assumed to your comment was on _total time_ complaint (which does not exist).   So I was wrong here -- my apologies.   It's still somewhat unrelated to data.table and thread count issue in a package though.

**Comment by aitap** [comment]
Does CRAN still set `options(Ncpus=2)`? If that's the case, we can at least postpone this by a few months by letting [`initDTthreads`](https://github.com/Rdatatable/data.table/blob/88039186915028ab3c93ccfd8e22c0d1c3534b1a/src/openmp-utils.c#L32) know the value of `getOption('Ncpus')` to use as an additional limit.

**Comment by reedacartwright** [comment]
It looks like `parallelly::availableCores()` handles a majority of the cases that people have brought up such as detecting whether the code is running on a desktop vs an HPC. This function (or an equivalent one) does seem to offer a way to define a default number of threads that would scale across systems.

https://parallelly.futureverse.org/reference/availableCores.html

**Comment by cdriveraus** [comment]
Did I miss anywhere in this thread where a simple solution was posted? One that handles examples as well as tests etc? Modifying every example on so many packages because CRAN removed an environment variable seems a little wild...

**Comment by cdriveraus** [comment]
Does anyone know why this doesn't work?

```
.onLoad <- function(libname,pkgname){ #set omp and dt thread limit on CRAN
  if(identical(Sys.getenv("NOT_CRAN"), "false")){
    data.table::setDTthreads(2)
    Sys.setenv("OMP_THREAD_LIMIT" = 2)
  }
}
```

**Comment by eddelbuettel** [comment]
> Does anyone know why this doesn't work?

Where does it work, or not work?  Who is supposed to set `NOT_CRAN`?  

CRAN doesn't so your boolean expression is FALSE.  Hence your expression body is not entered.

**Comment by jangorecki** [comment]
Looking at discussions like https://github.com/microsoft/LightGBM/issues/5987 and https://github.com/dmlc/xgboost/issues/9497#issuecomment-1717557887 it is really surprising that there is so much resistance towards simply setting an env var on CRAN. Maybe there is some good reason that we just haven't been told? Otherwise it seems insane to generate days or weeks of developer hours across whole R ecosystem. And not just that, when you see comments like

> Alternatively, we just restrict the number of threads during model load for R build.

then it's clear it is not healthy for R ecosystem. People will notice that xgboost is faster on python and just move there.
Maybe entities like R consortium/similar could be involved to find a solution.

**Comment by MichaelChirico** [comment]
> we just haven't been told

Linking (without endorsement) Uwe's reasoning from the r-devel thread:

https://stat.ethz.ch/pipermail/r-package-devel/2023q3/009463.html

https://stat.ethz.ch/pipermail/r-package-devel/2023q3/009466.html

**Comment by eddelbuettel** [comment]
Yes.  We tried.  I had a redunkiously high number of posts in those threads, but all to no avail.  "Their" decision is wrong, will lead to numerous rounds of revision by package authors who each have to learn the 'how to do this' one by one, and (as Jan noted above) may well make R packages look inferior to non-R variants.  It sucketh.

**Comment by jangorecki** [comment]
> > we just haven't been told
> 
> Linking (without endorsement) Uwe's reasoning from the r-devel thread:
> 
> https://stat.ethz.ch/pipermail/r-package-devel/2023q3/009463.html
> 
> https://stat.ethz.ch/pipermail/r-package-devel/2023q3/009466.html

If that would be the reason then I don't understand why they want R packages to be updated rather than just setting this env var for every R session, not only on CRAN. Wouldn't that achieve what Uwe is telling about? in a proper place? One change in one place and thousands of R packages fixed.

Having that is of course nonsense considering modern computer architecture.

**Comment by jangorecki** [comment]
Because most of people who reach here are just looking to fix their packages. Let's reduce the burden caused by this change on CRAN, provide workable solution and close this issue, where there is nothing else to do than just discussing about the craziness of the situation.

I haven't tested it, so please let me know if it won't work.

- to fix tests just use `data.table::setDTthreads(2)` at the start of your test script:
```r
data.table::setDTthreads(2)
library(my_package)
stopifnot(my_function(rnorn(5)))
```

- to fix examples use `\dontshow{data.table::setDTthreads(2)}`:
```Rd
\examples{
  \dontshow{data.table::setDTthreads(2)}
  my_function(rnorm(5))
}
```
see description of `examples` tag on https://cran.r-project.org/doc/manuals/R-exts.html#Documenting-functions for explanation if not clear.

- to fix vignettes use knitr chunk option `echo`
````Rmd
```{r chunkname, echo=-1}
data.table::setDTthreads(2)
my_function(rnorn(5))
```
````

That should work for data.table rev deps. If you are package developer that uses OpenMP directly then I am afraid you have to provide a mechanism like `setDTthreads` yourself and apply as described here.

**Comment by aitap** [comment]
In addition to what @jangorecki says:

If you're using `parallel`-style clusters together with `data.table`, you need to split your core allowance between the two. This means either `makeCluster(1)` and `setDTthreads(2)` or the other way around. (This problem isn't novel or unique to `data.table`. Anyone with OpenBLAS/MKL/other parallel BLAS and a `parallel` cluster has already had to manage it somehow.)

It's best to minimise changes to global state from the examples. Start your examples with `\dontshow{.prev.dt.threads <- data.table::setDTthreads(2)}` *and* end them with `\dontshow{data.table::setDTthreads(.prev.dt.threads); rm(.prev.dt.threads)}`. (Use a variable name starting with a dot because those are by convention not for the user to care about despite being present in the global environment.) Otherwise a user that had set up their DT threads to their liking will be surprised to see `data.table` slowed down again after running `example(your_function)`.

**Comment by EmilHvitfeldt** [comment]
@jangorecki, your suggested solutions does not work in my use case.

with https://github.com/tidymodels/textrecipes/pull/251 I still get the CRAN submission NOTE with

```md
* checking examples ... [17s/12s] NOTE
Examples with CPU time > 2.5 times elapsed time
                 user system elapsed ratio
step_dummy_hash 1.579   0.11   0.273 6.187
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ... [24s/19s] OK
  Running ‘testthat.R’ [24s/18s]
```

I don't know if it is because I don't use {data.table} directly and instead using a package {text2vec} that uses {data.table}.

Last release I ended up simply deleting the documentation for the affected functions, but that is not a long-term solution.

**Comment by jangorecki** [comment]
@EmilHvitfeldt hard for me to imagine why that could still cause problem, unless there is some nested parallelism, like mclapply + data.table, then inside topmost parallel code should be call to set any nested code to be single threaded, not just for data.table but for any multi threaded computations. Are you able to narrow down from where parallel processing comes?

**Comment by eddelbuettel** [comment]
It again demonstrates that CRAN was  not really helpful here by insisting each and every package fix that on their own.

**Comment by cdriveraus** [comment]
given that they are considering a general solution to the problem, waiting for that before rigidly enforcing the 2 core thing would have seemed sensible to me...

**Comment by eddelbuettel** [comment]
@cdriveraus News to me.  Can you supply a reference to back up that clain? Last we all saw was Uwe Ligges telling everybody to set up 'max 2 cores' in each package needing it.

**Comment by cdriveraus** [comment]
@eddelbuettel I thought it was mentioned somewhere on the pkg devel list at one point, but I'm guessing you're more familiar than me so I'm assuming I misinterpreted / imagined one of the posts I read...

**Comment by EmilHvitfeldt** [comment]
One of the frustration part of this problem is that I'm not able to reproduce this problem locally. Unless broken, the {text2vec} parallelism is only for certain functions which I'm not using (as far as I know).

Only seeing it on the Debian incoming check machine.

right now my examples have the following to no avail. 

```r
#' \dontshow{library(data.table)}
#' \dontshow{data.table::setDTthreads(2)}
#' \dontshow{Sys.setenv("OMP_THREAD_LIMIT" = 2)}
#' \dontshow{library(text2vec)}
#' \dontshow{options("text2vec.mc.cores" = 1)}
```

**Comment by helske** [comment]
I think there is something else going on on CRAN than just the `data.table` behaviour. I have similar problems with the [`bssm`](https://github.com/helske/bssm) package, which does not use `data.table`, but uses OpenMP in some selected functions where the number of threads is set by the the user in the R function call (default being one). I tried using `Sys.setenv("OMP_THREAD_LIMIT" = 2)` at the start of the problematic example, but I still get the same pretest check note from CRAN on Debian.

**Comment by jangorecki** [comment]
yes, problem is not with data.table but openmp (or other multithreading code).

it is interesting that even in your case where it uses single thread by default, it still having the problem. Then I don't think there is any reliable solution that package maintainers can employ.

**Comment by stitam** [comment]
Not sure this a proper solution, but this is how I handle the issue within the [webseq](https://github.com/stitam/webseq) package. The package is not yet on CRAN but it does pass `devtools::check()` with the default `cran = TRUE`.

In each of my functions that use parallelisation, I set the default value for the number of cores to `NULL` and validate the number of cores with an internal `get_mc_cores()` function which 1. returns the number of cores if it has been set, 2. otherwise looks at `getOption("Ncpu")` and 3. if that returns `NULL`, falls back to `parallel::detectCores()`. 

Within a regular R session I do not set "Ncpu" so R will not find it and will fall back to using many cores by default. However, I start each of my test files with `options("Ncpu" = 2L)` so R CMD check will always work with 2 cores.

Do you folks think this will work on CRAN? Thanks.

**Comment by jangorecki** [comment]
DT uses srtDTthreads rather than global options. Please read manual.

**Comment by aitap** [comment]
R CMD check also runs example() for all your help pages, so any
examples that use parallel processing must pass mc_cores = at most 2.
Do not set options(Ncpu=2L) in examples because the user may have set
it to a different value they are more comfortable with, and running
such an example would change user's observable global state.

What will your code do if parallel::detectCores() returns NA?

**Comment by stitam** [comment]
Thanks @aitap! Oh, my examples were wrapped in `dontrun{}` so they were not evaluated by R CMD check.. I unwrapped them to see what happens and R CMD check threw an error, as expected. I'm okay with wrapping, and AFAIK it is not against CRAN policies, but if that's not acceptable, then unfortunately setting `options("Ncpu" = 2L)` within tests is not enough to pass the checks. Thanks for flagging the `NA`, currently nothing. I am unsure exactly when `parallel::detectCores()` returns `NA`, so I'll probably just add `stop()` for now.

**Comment by aadler** [comment]
You may want to look into the [parallelly](https://github.com/HenrikBengtsson/parallelly) package's [`availableCores`](https://github.com/HenrikBengtsson/parallelly) function which is written to always return an integer. Increases your dependency count but may be worth it.

**Comment by grantmcdermott** [comment]
Sorry to dredge up an old thread, but I've just been bitten by this again with a new CRAN submission.

Reading through the thread, I'm not sure whether there is a truly bullet proof solution. But I think @jangorecki's suggestions in https://github.com/Rdatatable/data.table/issues/5658#issuecomment-1741934995 seem the most sensible port of first call, alongside the `Sys.setenv("OMP_THREAD_LIMIT" = 2)` option mentioned by @lrberge, @eddelbuettel and others.

Should these suggestions not be added to the [Importing data.table](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-importing.html) vignette?

**Comment by eddelbuettel** [comment]
It amazes me that this is still being discussed. Some of us fixed it years ago. It can be done zero-dependency too.  YMMV.

**Comment by aitap** [comment]
Unless some downstream code deliberately takes a look at the
environment variables and limits its own thread usage (as suggested in
<https://github.com/Rdatatable/data.table/issues/5658#issuecomment-1592264899>),
Sys.setenv("OMP_THREAD_LIMIT" = 2) may only work for the child
processes, not OpenMP threads spawned by the current process. The
OpenMP runtime is required to ignore Sys.setenv() done by the current
process:

> Modifications to the environment variables after the program has
> started, even if modified by the program itself, are ignored by the
> OpenMP implementation.

https://www.openmp.org/spec-html/5.0/openmpch6.html#x287-20510006

**Comment by aadler** [comment]
For a while already I've followed a form of @eddelbuettel advice and basically just set the cores to 2 before every test ([like so](https://github.com/aadler/Delaporte/blob/master/inst/tinytest/test-ddelap.r)). Tests are almost never run by users, only developers and CRAN, so it should not affect performance.
